### DISCUSSION TITLE
Automatic build number is always set to 1

### DISCUSSION THREAD

**Discussion by mostafaAbdelazim** [codemagic-ci-cd/codemagic-docs#2251] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

Add this script to prebuild script

  scripts:
    - name: Get the latest build number
      script: |
        LATEST_BUILD_NUMBER=$(app-store-connect get-latest-testflight-build-number $APP_APPLE_ID)
        cd ./ios
        xcrun agvtool new-version -all $(($LATEST_BUILD_NUMBER + 1))

### Expected results

i would expect the new build number to be LATEST_BUILD_NUMBER + 1

### Actual results

Found build number 4 from TestFlight version 1.0.2
Setting version of project Runner to: 
    1.

Also setting CFBundleVersion key (assuming it exists)

Updating CFBundleVersion in Info.plist(s)...

Updated CFBundleVersion in "Runner.xcodeproj/../Runner/Info.plist" to 1

### Build id (optional)

65717eecbe7b02f1e859a4e3

**Comment by dtrdic** [comment, marked as answer]
Hey @mostafaAbdelazim, you need to add the logic inside build arguments and not in the pre-build script, something like this
`--build-number=$(($(app-store-connect get-latest-testflight-build-number "$APP_ID") + 1))`

you can also use a built-in variables like `PROJECT_BUILD_NUMBER` or `BUILD_NUMBER`

`--build-name=1.2.$PROJECT_BUILD_NUMBER --build-number=$(($PROJECT_BUILD_NUMBER + 1))`

![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/38918425/a1998029-d77d-4a73-a64b-988f94aa395f)


The `PROJECT_BUILD_NUMBER` variable is a built-in environment variable that represents the current build number of your project.
It is automatically incremented with each new build of your app.
image

Also make sure that Info.plist file has the following string along with their key values

```
<key>CFBundleShortVersionString</key>
<string>$(FLUTTER_BUILD_NAME)</string>
<key>CFBundleVersion</key>
<string>$(FLUTTER_BUILD_NUMBER)</string>
```

**Reply by mostafaAbdelazim** [reply]
Working with PROJECT_BUILD_NUMBER seems to be enough for the task, Thank you @dtrdic,
Is the official docs outdated ? I followed this [section](https://docs.codemagic.io/knowledge-codemagic/build-versioning/#app-store-or-testflight-latest-build-number) with the pre-build script

**Reply by dtrdic** [reply]
awesome @mostafaAbdelazim, i'm glad the issue has been resolved!
No, it's still valid, but for the yaml configuration, I agree we should clarify a little bit better in the docs, I'll make sure to update it accordingly..Cheers
